enhancements/microshift/align-etcd-directories-in-microshift.md

```diff
@@ -0,0 +1,147 @@
+authors:
+  - dusk125
+reviewers:
+  - "@hasbro17, etcd Team"
+  - "@tjungblu, etcd Team"
+  - "@Elbehery, etcd Team"
+  - "@fzdarsky"
+  - "@derekwaynecarr"
+  - "@mangelajo"
+  - "@pmtk"
+approvers:
+  - "@dhellmann"
+  - "@deads2k"
+api-approvers:
+  - None
+creation-date: 2023-02-20
+last-updated: 2023-02-20
+tracking-link:
+  - https://issues.redhat.com/browse/ETCD-356
+---
+
+# Align etcd directories in Microshift
+
+## Summary
+
+In Openshift, etcd and related tools (like etcdctl and sos) rely on the directories `/var/lib/etcd`, `/etc/kubernetes/static-pod-certs`, `/etc/kubenetes/static-pod-resources` to exist and to contain relevant files by default. In Microshift, all of the Microshift related information (including etcd) is located under `/var/lib/microshift` (with the etcd data directory `/var/lib/microshift/etcd`).
+
+## Motivation
+
+This enhancement opens a discussion for deciding how best to align, if at all, the above directories. 
+The three proposed options are:
+
+1. To do no alignment and keep all etcd directories under `/var/lib/microshift`.
+2. Symlink the etcd directories under `/var/lib/microshift` to their Openshift counterparts (listed above).
+3. Update Microshift to use the Openshift-expected locations.
+
+### User Stories
+
+TODO
+
+### Goals
+
+- Select one of the approaches to implement.
+
+### Non-Goals
+
+TODO
+
+## Proposal
+
+### No Alignment
+With this option, Microshift is not updated and everything (etcd related) remains under the `/var/lib/microshift` base directory; this includes the etcd data directory, certs directory, and the scripts directory. 
+Selecting this option will essentially have Microshift go down the same path as Hypershift in that the Microshift team will be responsible for owning the delivery, documentation, support, scripting, etc of etcd when those differ from those of Openshift. The etcd team would still assist for lower-level etcd issues: those that pertain to the server itself and not its deployment.
+
+### Symlink Alignment
+This is the middle ground between No Alignment and Full Alignment; the current proposal is to keep everything stored under `/var/lib/microshift` and symlink its contents to the Openshift directories.
+This would allow the Openshift documentation, etc to still apply to Microshift; however, there would need to validation done to ensure that scripts and workflows will work appropiately through a symlink.
+
+### Full Alignment
+This approach would include an update to Microshift that would make the etcd data, scripts, and certs directories match that of Openshift so that all documentation, scripts, and knowledge would match one-for-one in Microshift and Openshift.
+This would allow the etcd product team to provide full support for Microshift, as it does for Openshift.
+
+### Workflow Description
+
+TODO
+
+#### Variation [optional]
+
+TODO
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+
+TODO
+
+### Risks and Mitigations
```

Comment:
I've seen an etcd RPM in brew. Is that shipped in RHEL? Do we have to worry about collisions with the etcd in MicroShift if the service from that RPM is installed alongside MicroShift? (Granted, that doesn't seem especially likely, but it would be pretty challenging to debug if it did happen.)

Reply:
According to [package browser](https://access.redhat.com/downloads/content/package-browser) `etcd` is present in OpenStack repositories.

Etcd config shipped there listens on `localhost:23{79,80}` which matches what we'll use after Pablo's PR (removal of IPs in certs). So in that scenario I'd say microshift will fail to run its etcd because ports are already bound

Reply:
If we didn't align the data directory, then we could easily handle this case by changing the ports used by microshift-etcd to something non-default.

Reply:
Do we know anything about how that version of etcd is selected or who maintains the package? Is there any chance that we want to reuse it instead of packaging our own?

If we don't want to reuse it, we need to decide if we want to shift the MicroShift settings off of the defaults to allow that other package to be installed, or if we would rather say that the microshift RPM conflicts with the etcd RPM so they can't both be installed at the same time.

Reply:
> If we didn't align the data directory, then we could easily handle this case by changing the ports used by microshift-etcd to something non-default.

That would require recompiling the microshift-etcd, we hardcoded the ports

Reply:
Yeah we could hardcode the ports to be whatever we want to avoid a potential conflict.

Reply:
I'm not even sure if all subscriptions would allow to install it from that repo, openstack repo has other stuff like ansible, puppet, collectd, and obviously openstack. Would average OpenShift/RHEL subscriber have access to that?

`etcd` RPM is just one way to install it. etcd's doc refer to prebuilt binaries or building from source, package manager ways are "TBD".

Also, if I'd run k8s cluster on a device, then I'd probably also deploy etcd (for my apps) on the cluster - but you never know someone else's usecase :)

Reply:
@pmtk those are good points. I suspect we aren't going to want to reuse the RPM. I don't know how to judge the risk of whether someone would have access to the OpenStack repos and pull in the etcd RPM accidentally.

Maybe the risk is low enough that we don't worry about it for now? We can change our hard-coded ports later, right?

Reply:
Yup, I'd stick with what we have now and document ports MicroShift is using on OS level (6443, 2379, 2380, 80 & 443 (router),  5353 (mDNS), and others) so users/system {admins,integrators} are aware of the requirements

Reply:
Yeah, we could rename this one from "aligning etcd data dir" to aligning "microshift-etcd with openshift / upstream etcd" and record our thoughts from comments. @dusk125 what do you think?

Reply:
>  Do we have to worry about collisions with the etcd in MicroShift if the service from that RPM is installed alongside MicroShift? 

If you install the etcd package, it will also install the systemd service. You can mask that service, but it would be good to have it documented somewhere, if this package will be installed, do `systemctl mask etcd` . We install this on 3.11 (even when etcd runs as static pod) to obtain the `etcdctl`. One of the steps is to disable the service.

We can add a exclude in the `yum.conf` to never install etcd package even if someone attempts to do that.